Deserialize and deconstruct `Notify` method call arguments in one go.

luukvanderduim opened the pull request:
Your question in chat was how to deserialize the 'actions:` field directly into a `Vec<Actions>`.
We unfortunately cannot because the signature says the field is 'as' (Array of Strings).
Nothing in that description guarantees complete actions. Would it have said 'a(ss)', we could have.

This commit does the next best thing:

- Remove `RawNotifySignature<'a>` as we no longer need it.

- Adjusts `Action` fields to resemble with what specs say about `Notify.actions` method call arguments.

	See: [Notification specifications](https://specifications.freedesktop.org/notification-spec/notification-spec-latest.html#commands):
	Actions are sent over as a list of pairs: the identifier for the
	action and he localized string that will be displayed to the user.

Before the `Actions` fields (name, method) were possibly meant 
for the invocation of the `Action` (?).

- Introduce `NotifyArgs` struct that holds and names the arguments to the `Notify` method call we are monitoring for.
- Make sure to derive `Type` + `Deserialize` on `NotifyArgs`.
 - Adjust `TryFrom<Message> for Notification` to simplify deserialisation and deconstruction of the `Message::body`.

 Since we have `NotifyArgs` struct and it has the `Type + Deserialize` traits, we can now simply

 ```Rust
 let notify_args = msg.body::<NotifyArgs>body()?;
 ```

 since we are interested in just a few fields, we can selectively
 deconstruct in one go:

 ```Rust
 let NotifyArgs { app_name, summary, body, actions, .. } = msg.body()?;
```

(- Format lib.rs)

Just take whatever you find useful.

luukvanderduim committed:
Deserialize `Notify` arguments

- Adjusts `Action` fields to correspond with what I understand specs say about
`Notify` method call arguments.

	See: [Notification specifications]():
	Actions are sent over as a list of pairs. [...] the identifier for the
	action and he localized string that will be displayed to the user.

	Previous `Actions` fields (name, method) are meant for the invocation
	of the `Action` (?).

- Remove `RawNotifySignature<'a>`
- Introduce `NotifyArgs` struct that holds and names the arguments to
the `Notify` method call we are monitoring for.
- Make sure to derive `Type` on `NotifyArgs`.
 - Adjust `TryFrom<Message> for Notification` to
 simplify deserialize and deconstruct the `Message::body`.

 Since we have the `NotifyArgs` struct and `NotifyArgs` has the
 `Type + Deserialize` traits, we can now simply

 ```Rust
 let notify_args = msg.body::<NotifyArgs>body()?;
 ```

 and since we are interested in just a few fields, we can selectively
 decunstruct in one go:

 ```Rust
 let NotifyArgs { app_name, summary, body, actions, .. } = msg.body()?;
```

(- Format lib.rs)

Test passes. Cherry-pick whatever you like.

Seems to work.

albertotirla commented:
thank you for your contribution, it's highly appreciated! Before I merge this though, there's one thing to consider:

* this is not ment as a notification server, the notifications are not ment, at least at the time of writing, to be viewed in any such program where all the fields of the specification are required. I probably should not have included actions, but I did because it seemned relatively easy to do, at the time anyway. This is ment to eventually be included in odilia, where the only arguments I would think pertinent are app name, summary and body, because that's all the user hears. To actually interact with the notification, odilia would have to be a notification server in order to be able to send the signal for when an action was selected, and if that would be the case, this crate would no longer be needed because we would have less hacky ways of ggoing about it
* optional: when you're extracting the action pairs, chunking by two and so on, perhaps explain in the comments more clearly why? or, perhaps only extract the localized versions, and then have actions be a newtype over string, which would be better in my opinion
Once this small issue is solved, this pr can be merged, because I see no other issues with it

luukvanderduim committed:
Addresses suggestions:

Make things lean.

- Removes `Action`s altogether.
- Remove Notify,
- Simplifies `Notification`
- Simplifies `TryFrom<Merssage> for Notification`
- Adjust test, since the actions field is no longer there.

The types we do not need are references in our signature in the hope that,
should Serde deserialize these into a temporary, these are zero-copy, borrowed
types before these are dropped.

luukvanderduim commented:
> not meant to be viewed in any such program where all the fields of the specification are required.

The struct is there to obtain the `Signature` via the Type  trait and be able to refer to these by name. 
If you have no interest in any of these, not even in future time, your approach was leaner.

Let me try to make it lean.

> I probably should not have included actions, but I did because it seemed relatively easy to do, at the time anyway.

Throw more out. Done!

albertotirla commented:
alright, it's much simpler and the API surface is better now.
If we would need the extra stuff for odilia itself in the future, the pr is still here and relevant, so your work was not in vane either way, as the code didn't disappear
Now this can be safely merged, and that is exactly what I will do.

albertotirla commented on src/action.rs:
this is no longer needed, deleted to simplify API
